lib/extras/enc/jpg.cc

```diff
@@ -235,5 +233,90 @@ Status EncodeImageJPG(const CodecInOut* io, JpegEncoder encoder, size_t quality,
   return true;
 }
 
+Status ParseChromaSubsampling(const std::string& param,
+                              YCbCrChromaSubsampling* subsampling) {
+  constexpr std::pair<const char*,
+                      std::pair<std::array<uint8_t, 3>, std::array<uint8_t, 3>>>
+      options[] = {{"444", {{{1, 1, 1}}, {{1, 1, 1}}}},
+                   {"420", {{{2, 1, 1}}, {{2, 1, 1}}}},
+                   {"422", {{{2, 1, 1}}, {{1, 1, 1}}}},
+                   {"440", {{{1, 1, 1}}, {{2, 1, 1}}}}};
+  for (const auto& option : options) {
+    if (param == option.first) {
+      return subsampling->Set(option.second.first.data(),
+                              option.second.second.data());
+    }
+  }
+  return false;
+}
+
+class JPEGEncoder : public Encoder {
+  std::vector<JxlPixelFormat> AcceptedFormats() const override {
+    std::vector<JxlPixelFormat> formats;
+    for (const uint32_t num_channels : {1, 3}) {
+      for (JxlEndianness endianness : {JXL_BIG_ENDIAN, JXL_LITTLE_ENDIAN}) {
+        formats.push_back(JxlPixelFormat{/*num_channels=*/num_channels,
+                                         /*data_type=*/JXL_TYPE_UINT8,
+                                         /*endianness=*/endianness,
+                                         /*align=*/0});
+      }
+    }
+    return formats;
+  }
+  Status Encode(const PackedPixelFile& ppf, EncodedImage* encoded_image,
+                ThreadPool* pool = nullptr) const override {
+    const auto& options = this->options();
+    int quality = 100;
+    auto it_quality = options.find("q");
+    if (it_quality != options.end()) {
+      std::istringstream is(it_quality->second);
+      JXL_RETURN_IF_ERROR(static_cast<bool>(is >> quality));
```

Comment:
Why not stoll?

Reply:
It appears to rely on exceptions for error handling: https://en.cppreference.com/w/cpp/string/basic_string/stol#Exceptions

Reply:
Fair enough. I guess we could strtoll, but that's not exactly very nice either...